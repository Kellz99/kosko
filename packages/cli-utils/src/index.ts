export { type GlobalArguments, globalOptions, parse } from "./command";
export { CLIError, handleError } from "./error";
export { setupLogger } from "./logger";
