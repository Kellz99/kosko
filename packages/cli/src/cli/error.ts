import exit from "exit";
import logger, { LogLevel } from "@kosko/log";

export class CLIError extends Error {
  public readonly name = "CLIError";
  public readonly output?: string;
  public readonly code?: number;

  public constructor(
    msg: string,
    { output, code }: { output?: string; code?: number } = {}
  ) {
    super(msg);
    this.output = output;
    this.code = code;
  }
}

export function handleError(err: unknown): void {
  let code = 1;

  if (err instanceof CLIError) {
    if (err.code != null) {
      code = err.code;
    }

    if (err.output) {
      logger.log(LogLevel.Error, err.output);
    } else {
      logger.log(LogLevel.Error, err.message, { error: err });
    }
  } else if (err instanceof Error) {
    if (err.name !== "YError") {
      logger.log(LogLevel.Error, "", { error: err });
    }
  } else {
    logger.log(LogLevel.Error, "", { error: err });
  }

  exit(code);
}
