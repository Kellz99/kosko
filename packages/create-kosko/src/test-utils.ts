import { join, posix } from "node:path";
import glob from "fast-glob";
import { readFile } from "node:fs/promises";

export async function listAllFiles(
  dir: string
): Promise<Record<string, string>> {
  const paths = await glob("**/*", { cwd: dir });
  const files: Record<string, string> = {};

  for (const path of paths) {
    files[posix.normalize(path)] = await readFile(join(dir, path), "utf8");
  }

  return files;
}
