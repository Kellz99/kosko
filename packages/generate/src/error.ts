import AggregateError from "@kosko/aggregate-error";
import extractStack from "extract-stack";

const STACK_INDENT = "    ";

interface Component {
  apiVersion: string;
  kind: string;
  metadata: {
    name: string;
    namespace?: string;
  };
}

function isComponent(value: unknown): value is Component {
  if (value == null || typeof value !== "object") return false;

  const { apiVersion, kind, metadata = {} } = value as any;

  return (
    typeof apiVersion === "string" &&
    typeof kind === "string" &&
    typeof metadata.name === "string"
  );
}

export interface ResolveErrorOptions {
  path?: string;
  index?: number[];
  cause?: Error;
  value?: unknown;
}

export function toError(value: unknown): Error {
  if (value instanceof Error) {
    return value;
  }

  if (typeof value === "string") {
    return new Error(value);
  }

  if (typeof value === "object" && value != null) {
    return Object.assign(new Error((value as any).message), value);
  }

  return new Error();
}

export function aggregateErrors(errors: Error[]) {
  if (errors.length === 1) {
    return errors[0];
  }

  return new AggregateError(errors);
}

function decorateErrorStack(
  err: Error,
  values: Record<string, string>
): string {
  const origStack = extractStack(err.stack);
  let stack = `${err.name}: ${err.message}`;

  for (const [key, value] of Object.entries(values)) {
    stack += `\n${STACK_INDENT}${key}: ${value}`;
  }

  if (origStack) stack += "\n" + origStack;

  return stack;
}

function generateCauseMessage(cause: Error): string {
  const causeStack = extractStack(cause);
  let msg = `${cause.name}: ${cause.message}`;

  if (causeStack) {
    msg +=
      "\n" +
      causeStack
        .split("\n")
        .map((line) => STACK_INDENT + line)
        .join("\n");
  }

  return msg;
}

export class ResolveError extends Error {
  public readonly path?: string;
  public readonly index?: number[];
  public readonly cause?: Error;
  public readonly value?: unknown;

  public constructor(message: string, options: ResolveErrorOptions = {}) {
    super(message);

    this.path = options.path;
    this.index = options.index;
    this.cause = options.cause;
    this.value = options.value;
    this.stack = decorateErrorStack(this, {
      ...(this.path && { Path: this.path }),
      ...(this.index?.length && { Index: `[${this.index.join(", ")}]` }),
      ...(isComponent(this.value) && {
        Kind: `${this.value.apiVersion}/${this.value.kind}`,
        ...(typeof this.value.metadata.namespace === "string" && {
          Namespace: this.value.metadata.namespace
        }),
        Name: this.value.metadata.name
      }),
      ...(this.cause && { Cause: generateCauseMessage(this.cause) })
    });
  }
}

ResolveError.prototype.name = "ResolveError";

export interface GenerateErrorOptions {
  path?: string;
  cause?: Error;
}

export class GenerateError extends Error {
  public readonly path?: string;
  public readonly cause?: Error;

  constructor(message: string, options: GenerateErrorOptions = {}) {
    super(message);

    this.path = options.path;
    this.cause = options.cause;
    this.stack = decorateErrorStack(this, {
      ...(this.path && { Path: this.path }),
      ...(this.cause && { Cause: generateCauseMessage(this.cause) })
    });
  }
}

GenerateError.prototype.name = "GenerateError";
