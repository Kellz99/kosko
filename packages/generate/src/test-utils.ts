/// <reference types="jest-extended" />
import { Manifest } from "./base";

type ManifestToMatch = Pick<Manifest, "data" | "metadata"> &
  Partial<Pick<Manifest, "position" | "issues">>;

export function matchManifest({
  position = { path: "", index: [] },
  issues = [],
  ...manifest
}: ManifestToMatch) {
  return {
    ...manifest,
    position,
    issues,
    report: expect.toBeFunction()
  };
}

export function matchManifests(manifests: ManifestToMatch[]) {
  return manifests.map(matchManifest);
}
