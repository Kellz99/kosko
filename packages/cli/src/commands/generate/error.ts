import { Issue, Manifest, Result, Severity } from "@kosko/generate";
import cleanStack from "clean-stack";
import pc from "picocolors";
import { CLIError } from "@kosko/cli-utils";
import { isRecord } from "@kosko/common-utils";
import { stderr } from "node:process";
import type { Formatter } from "picocolors/types";
import logger, { LogLevel } from "@kosko/log";

const severityFormats: Record<Severity, Formatter> = {
  error: pc.red,
  warning: pc.yellow
};

const severityIcons: Record<Severity, string> = {
  error: "✖",
  warning: "⚠"
};

function print(line: string): void {
  stderr.write(line + "\n");
}

interface ErrorLike {
  name: string;
  message: string;
  stack?: string;
}

function toErrorLike(err: unknown): ErrorLike | undefined {
  if (err instanceof Error) return err;

  if (isRecord(err)) {
    const { name, message, stack } = err;

    if (typeof message === "string") {
      return {
        name: (typeof name === "string" && name) || "Error",
        message,
        ...(typeof stack === "string" && { stack })
      };
    }
  }

  if (typeof err === "string") {
    return { name: "Error", message: err };
  }
}

export function resultHasError(result: Result): boolean {
  return result.manifests.some((manifest) =>
    manifest.issues.some((issue) => issue.severity === "error")
  );
}

function prettifyPath(cwd: string, path: string): string {
  if (path.startsWith(cwd)) {
    path = path.substring(cwd.length).replace(/^[/\\]+/, "");
  }

  return path.replace(/\\/g, "/");
}

interface IssueStats {
  error: number;
  warning: number;
}

function getIssueStats(manifest: Manifest): IssueStats {
  const stats: IssueStats = { error: 0, warning: 0 };

  for (const issue of manifest.issues) {
    stats[issue.severity]++;
  }

  return stats;
}

function printManifestHeader(cwd: string, manifest: Manifest): void {
  const { metadata, position } = manifest;
  const chunks: string[] = [];

  if (metadata) {
    chunks.push(`${metadata.apiVersion}/${metadata.kind}`);
    chunks.push(
      `${metadata.namespace ? metadata.namespace + "/" : ""}${metadata.name}`
    );
  }

  if (position.index.length) {
    chunks.push(`[${position.index.join(", ")}]`);
  }

  print(
    pc.underline(prettifyPath(cwd, position.path)) +
      (chunks.length ? ` - ${chunks.join(" ")}` : "")
  );
}

function printIssue(issue: Issue): void {
  const format = severityFormats[issue.severity];
  const icon = severityIcons[issue.severity];

  print(format(`${icon} ${issue.message}`));

  if (issue.cause) {
    const err = toErrorLike(issue.cause);
    if (!err) return;

    let stack = err.stack
      ? cleanStack(err.stack)
      : `${err.name}: ${err.message}`;

    stack = stack
      .split("\n")
      .map((line) => "  " + line)
      .join("\n");

    print(pc.gray(stack));
  }
}

function pluralize(count: number, word: string): string {
  return count === 1 ? word : word + "s";
}

function printSummary(stats: IssueStats): void {
  const chunks: string[] = [];

  if (stats.error) {
    chunks.push(`${stats.error} ${pluralize(stats.error, "error")}`);
  }

  if (stats.warning) {
    chunks.push(`${stats.warning} ${pluralize(stats.warning, "warning")}`);
  }

  if (!chunks.length) return;

  print("");

  logger.log(
    stats.error ? LogLevel.Error : LogLevel.Warn,
    `Found ${chunks.join(" and ")} in total`
  );
}

function createGenerateFailedError(): CLIError {
  return new CLIError("Generate failed", { output: "Generate failed" });
}

export function printIssues(cwd: string, result: Result): void {
  const totalStats: IssueStats = { error: 0, warning: 0 };

  for (const manifest of result.manifests) {
    if (!manifest.issues.length) continue;

    const stats = getIssueStats(manifest);
    totalStats.error += stats.error;
    totalStats.warning += stats.warning;

    print("");
    printManifestHeader(cwd, manifest);

    for (const issue of manifest.issues) {
      print("");
      printIssue(issue);
    }
  }

  printSummary(totalStats);

  if (totalStats.error) {
    throw createGenerateFailedError();
  }
}

function flattenError(err: unknown): unknown[] {
  if (err instanceof AggregateError) {
    return err.errors.flatMap(flattenError);
  }

  return [err];
}

export function handleGenerateError(err: unknown): void {
  const allErrors = flattenError(err);
  let count = 0;

  for (const value of allErrors) {
    const err = toErrorLike(value);
    if (!err) continue;

    count++;

    print("");
    print(
      severityFormats.error(
        `${severityIcons.error} ${err.name}: ${err.message}`
      )
    );

    if (err.stack) {
      const stack = cleanStack(err.stack).split("\n").slice(1).join("\n");

      print(pc.gray(stack));
    }
  }

  printSummary({ error: count, warning: 0 });
  throw createGenerateFailedError();
}
