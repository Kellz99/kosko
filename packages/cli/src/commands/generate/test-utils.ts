import { Manifest } from "@kosko/generate";

export function createManifest(
  value: Partial<Omit<Manifest, "report">> = {}
): Manifest {
  return {
    position: { path: "", index: [] },
    issues: [],
    data: undefined,
    report() {},
    ...value
  };
}
