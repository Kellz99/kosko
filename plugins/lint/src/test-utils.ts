import { Issue, Manifest, type ManifestToValidate } from "@kosko/generate";
import type { RuleContext, RuleFactory } from "./rules/types";
import assert from "node:assert";

export interface ReportedIssue {
  manifest: ManifestToValidate;
  message: string;
}

export interface TestContext<T> extends RuleContext<T> {
  issues: ReportedIssue[];
}

export function createTestContext<T>(config: T): TestContext<T> {
  const issues: ReportedIssue[] = [];

  return {
    config,
    severity: "error",
    issues,
    report(manifest, message) {
      issues.push({ manifest, message });
    }
  };
}

export function createManifest(
  value: unknown
): ManifestToValidate & Pick<Manifest, "issues"> {
  const issues: Issue[] = [];

  return {
    path: "",
    index: [],
    data: value,
    issues,
    report(issue) {
      issues.push(issue);
    }
  };
}

export function validate<T>(
  rule: RuleFactory<T>,
  config: T,
  manifest: ManifestToValidate
): ReportedIssue[] {
  const ctx = createTestContext(config);
  const { validate } = rule.factory(ctx);

  assert(typeof validate === "function");
  validate(manifest);

  return ctx.issues;
}

export function validateAll<T>(
  rule: RuleFactory<T>,
  config: T,
  manifests: readonly ManifestToValidate[]
): ReportedIssue[] {
  const ctx = createTestContext(config);
  const { validateAll } = rule.factory(ctx);

  assert(typeof validateAll === "function");
  validateAll(manifests);

  return ctx.issues;
}
