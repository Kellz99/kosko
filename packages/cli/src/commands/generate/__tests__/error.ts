/// <reference types="jest-extended" />
import { CLIError } from "@kosko/cli-utils";
import { handleGenerateError, printIssues } from "../error";
import BufferList from "bl";
import { createManifest } from "../test-utils";

let stderr: BufferList;

jest.spyOn(process.stderr, "write");

function createError(msg: string): Error {
  const err = new Error(msg);
  // Set a fake stack because Jest injects a lot of stack frames
  err.stack = `Error: ${msg}\n    at foo.ts:1:1`;
  return err;
}

beforeEach(() => {
  stderr = new BufferList();
  (process.stderr.write as jest.Mock).mockImplementation((chunk) => {
    stderr.write(chunk);
  });
});

describe("printIssues", () => {
  test("single error issue", () => {
    expect(() => {
      printIssues("", {
        manifests: [
          createManifest({
            position: { path: "components/foo.ts", index: [] },
            issues: [{ severity: "error", message: "Error message" }]
          })
        ]
      });
    }).toThrowWithMessage(CLIError, "Generate failed");

    expect(stderr.toString()).toMatchSnapshot();
  });

  test("single warning issue", () => {
    printIssues("", {
      manifests: [
        createManifest({
          position: { path: "components/foo.ts", index: [] },
          issues: [{ severity: "warning", message: "Warning message" }]
        })
      ]
    });

    expect(stderr.toString()).toMatchSnapshot();
  });

  test("manifest contains metadata", () => {
    expect(() => {
      printIssues("", {
        manifests: [
          createManifest({
            position: { path: "components/foo.ts", index: [] },
            issues: [{ severity: "error", message: "Error message" }],
            metadata: {
              apiVersion: "v1",
              kind: "Pod",
              name: "foo"
            }
          })
        ]
      });
    }).toThrowWithMessage(CLIError, "Generate failed");

    expect(stderr.toString()).toMatchSnapshot();
  });

  test("metadata contains namespace", () => {
    expect(() => {
      printIssues("", {
        manifests: [
          createManifest({
            position: { path: "components/foo.ts", index: [] },
            issues: [{ severity: "error", message: "Error message" }],
            metadata: {
              apiVersion: "v1",
              kind: "Pod",
              namespace: "default",
              name: "foo"
            }
          })
        ]
      });
    }).toThrowWithMessage(CLIError, "Generate failed");

    expect(stderr.toString()).toMatchSnapshot();
  });

  test("index is not empty", () => {
    expect(() => {
      printIssues("", {
        manifests: [
          createManifest({
            position: { path: "components/foo.ts", index: [2, 4] },
            issues: [{ severity: "error", message: "Error message" }]
          })
        ]
      });
    }).toThrowWithMessage(CLIError, "Generate failed");

    expect(stderr.toString()).toMatchSnapshot();
  });

  test("metadata and index are present", () => {
    expect(() => {
      printIssues("", {
        manifests: [
          createManifest({
            position: { path: "components/foo.ts", index: [2, 4] },
            issues: [{ severity: "error", message: "Error message" }],
            metadata: {
              apiVersion: "v1",
              kind: "Pod",
              name: "foo"
            }
          })
        ]
      });
    }).toThrowWithMessage(CLIError, "Generate failed");

    expect(stderr.toString()).toMatchSnapshot();
  });

  test("ignore manifests without issues", () => {
    expect(() => {
      printIssues("", {
        manifests: [
          createManifest(),
          createManifest({
            position: { path: "components/bar.ts", index: [] },
            issues: [{ severity: "error", message: "Error message" }]
          })
        ]
      });
    }).toThrowWithMessage(CLIError, "Generate failed");

    expect(stderr.toString()).toMatchSnapshot();
  });

  test("issue contains a string cause", () => {
    expect(() => {
      printIssues("", {
        manifests: [
          createManifest({
            position: { path: "components/foo.ts", index: [] },
            issues: [
              {
                severity: "error",
                message: "Error message",
                cause: "Cause message"
              }
            ]
          })
        ]
      });
    }).toThrowWithMessage(CLIError, "Generate failed");

    expect(stderr.toString()).toMatchSnapshot();
  });

  test("issue contains an object cause", () => {
    expect(() => {
      printIssues("", {
        manifests: [
          createManifest({
            position: { path: "components/foo.ts", index: [] },
            issues: [
              {
                severity: "error",
                message: "Error message",
                cause: {
                  name: "FooError",
                  message: "foobar"
                }
              }
            ]
          })
        ]
      });
    }).toThrowWithMessage(CLIError, "Generate failed");

    expect(stderr.toString()).toMatchSnapshot();
  });

  test("issue contains a Error class cause", () => {
    expect(() => {
      printIssues;
      printIssues("", {
        manifests: [
          createManifest({
            position: { path: "components/foo.ts", index: [] },
            issues: [
              {
                severity: "error",
                message: "Error message",
                cause: createError("foobar")
              }
            ]
          })
        ]
      });
    }).toThrowWithMessage(CLIError, "Generate failed");

    expect(stderr.toString()).toMatchSnapshot();
  });

  test("cause is a empty object", () => {
    expect(() => {
      printIssues("", {
        manifests: [
          createManifest({
            position: { path: "components/foo.ts", index: [] },
            issues: [
              {
                severity: "error",
                message: "Error message",
                cause: {}
              }
            ]
          })
        ]
      });
    }).toThrowWithMessage(CLIError, "Generate failed");

    expect(stderr.toString()).toMatchSnapshot();
  });

  test("multiple issues", () => {
    expect(() => {
      printIssues("", {
        manifests: [
          createManifest({
            position: { path: "components/foo.ts", index: [] },
            issues: [
              { severity: "error", message: "Error 1" },
              { severity: "warning", message: "Warning 1" },
              { severity: "error", message: "Error 2" }
            ]
          }),
          createManifest({
            position: { path: "components/bar.ts", index: [] },
            issues: [{ severity: "error", message: "Error 3" }]
          })
        ]
      });
    }).toThrowWithMessage(CLIError, "Generate failed");

    expect(stderr.toString()).toMatchSnapshot();
  });

  test("cwd is not empty", () => {
    expect(() => {
      printIssues("/foo/bar", {
        manifests: [
          createManifest({
            position: { path: "/foo/bar/components/foo.ts", index: [] },
            issues: [{ severity: "error", message: "Error message" }]
          }),
          createManifest({
            position: { path: "/abc/def/components/bar.ts", index: [] },
            issues: [{ severity: "error", message: "Error message" }]
          })
        ]
      });
    }).toThrowWithMessage(CLIError, "Generate failed");

    expect(stderr.toString()).toMatchSnapshot();
  });
});

describe("handleGenerateError", () => {
  test("normal error", () => {
    expect(() => {
      handleGenerateError(createError("foo"));
    }).toThrowWithMessage(CLIError, "Generate failed");

    expect(stderr.toString()).toMatchSnapshot();
  });

  test("AggregateError", () => {
    expect(() => {
      handleGenerateError(
        new AggregateError([createError("foo"), createError("bar")])
      );
    }).toThrowWithMessage(CLIError, "Generate failed");

    expect(stderr.toString()).toMatchSnapshot();
  });

  test("nested AggregateError", () => {
    expect(() => {
      handleGenerateError(
        new AggregateError([
          createError("foo"),
          new AggregateError([createError("bar"), createError("baz")])
        ])
      );
    }).toThrowWithMessage(CLIError, "Generate failed");

    expect(stderr.toString()).toMatchSnapshot();
  });

  test("value is undefined", () => {
    expect(() => {
      handleGenerateError(undefined);
    }).toThrowWithMessage(CLIError, "Generate failed");

    expect(stderr.toString()).toBeEmpty();
  });
});
