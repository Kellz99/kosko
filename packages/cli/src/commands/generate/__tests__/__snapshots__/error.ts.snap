// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleGenerateError AggregateError 1`] = `
"
[31mâœ– Error: foo[39m
[90m    at foo.ts:1:1[39m

[31mâœ– Error: bar[39m
[90m    at foo.ts:1:1[39m

[31merror[39m - Found 2 errors in total
"
`;

exports[`handleGenerateError nested AggregateError 1`] = `
"
[31mâœ– Error: foo[39m
[90m    at foo.ts:1:1[39m

[31mâœ– Error: bar[39m
[90m    at foo.ts:1:1[39m

[31mâœ– Error: baz[39m
[90m    at foo.ts:1:1[39m

[31merror[39m - Found 3 errors in total
"
`;

exports[`handleGenerateError normal error 1`] = `
"
[31mâœ– Error: foo[39m
[90m    at foo.ts:1:1[39m

[31merror[39m - Found 1 error in total
"
`;

exports[`printIssues cause is a empty object 1`] = `
"
[4mcomponents/foo.ts[24m

[31mâœ– Error message[39m

[31merror[39m - Found 1 error in total
"
`;

exports[`printIssues cwd is not empty 1`] = `
"
[4mcomponents/foo.ts[24m

[31mâœ– Error message[39m

[4m/abc/def/components/bar.ts[24m

[31mâœ– Error message[39m

[31merror[39m - Found 2 errors in total
"
`;

exports[`printIssues ignore manifests without issues 1`] = `
"
[4mcomponents/bar.ts[24m

[31mâœ– Error message[39m

[31merror[39m - Found 1 error in total
"
`;

exports[`printIssues index is not empty 1`] = `
"
[4mcomponents/foo.ts[24m - [2, 4]

[31mâœ– Error message[39m

[31merror[39m - Found 1 error in total
"
`;

exports[`printIssues issue contains a Error class cause 1`] = `
"
[4mcomponents/foo.ts[24m

[31mâœ– Error message[39m
[90m  Error: foobar
      at foo.ts:1:1[39m

[31merror[39m - Found 1 error in total
"
`;

exports[`printIssues issue contains a string cause 1`] = `
"
[4mcomponents/foo.ts[24m

[31mâœ– Error message[39m
[90m  Error: Cause message[39m

[31merror[39m - Found 1 error in total
"
`;

exports[`printIssues issue contains an object cause 1`] = `
"
[4mcomponents/foo.ts[24m

[31mâœ– Error message[39m
[90m  FooError: foobar[39m

[31merror[39m - Found 1 error in total
"
`;

exports[`printIssues manifest contains metadata 1`] = `
"
[4mcomponents/foo.ts[24m - v1/Pod foo

[31mâœ– Error message[39m

[31merror[39m - Found 1 error in total
"
`;

exports[`printIssues metadata and index are present 1`] = `
"
[4mcomponents/foo.ts[24m - v1/Pod foo [2, 4]

[31mâœ– Error message[39m

[31merror[39m - Found 1 error in total
"
`;

exports[`printIssues metadata contains namespace 1`] = `
"
[4mcomponents/foo.ts[24m - v1/Pod default/foo

[31mâœ– Error message[39m

[31merror[39m - Found 1 error in total
"
`;

exports[`printIssues multiple issues 1`] = `
"
[4mcomponents/foo.ts[24m

[31mâœ– Error 1[39m

[33mâš  Warning 1[39m

[31mâœ– Error 2[39m

[4mcomponents/bar.ts[24m

[31mâœ– Error 3[39m

[31merror[39m - Found 3 errors and 1 warning in total
"
`;

exports[`printIssues single error issue 1`] = `
"
[4mcomponents/foo.ts[24m

[31mâœ– Error message[39m

[31merror[39m - Found 1 error in total
"
`;

exports[`printIssues single warning issue 1`] = `
"
[4mcomponents/foo.ts[24m

[33mâš  Warning message[39m

[33mwarn [39m - Found 1 warning in total
"
`;
